### ISSUE TITLE
Seat heater architecture plans

### ISSUE THREAD

**Issue by mattsch** [zabuldon/teslajsonpy#192]
After poking around a bit at the new seat heater functionality (thanks @alandtse and @Trupal00p), I starting thinking about current behavior and how it will integrate into HA.

Firstly, the current approach of checking for availability at startup can't work since it doesn't have the data for the car until after the initial setup and then **only** if the car is awake. My thought is that it should either add those as disabled entities by default and/or have a config option in the integration to enable them. I have a proof of concept on a local branch for the former option.

The other (probably more controversial option) is that we shouldn't actually represent them as switches but as sensors and then expose a service (tesla.set_heated_seats?) that controls them. There's no entity type that accurately represents the functionality so anything we do will be a hack. Fans might work as long as they can only have the `SUPPORT_PRESET_MODE` feature, and climate could work if we only set the `SUPPORT_PRESET_MODE` feature, but either one is a bit dirty IMO. I'm happy to work on an implementation for the service if that would be helpful as well.

**Comment by alandtse** [comment]
Do any cars have a seat warmer entity? Should we propose an entity for that?

While it's fine to just choose something, we have the opportunity to lead on this. Perhaps we need to reach out to the other car integrations.

On a general service, I'm tempted to opening a service call that allows accessing everything in the api. Someone on the forum wants to adjust the standard charging limit and that would require such a service.

**Comment by mattsch** [comment]
I haven't seen any mention of it, I'm sure other cars do but I don't see any other integrations that expose the functionality. I do like how the [BMW](https://www.home-assistant.io/integrations/bmw_connected_drive/#services) integration has services for the various functionality. Makes it very easy to see what it can do, and less error-prone on the user side. If we followed that same approach we could expose quite a bit more of the native commands that also don't have good translations to entity types.

If we wanted to go with entities, there is the [number](https://developers.home-assistant.io/docs/core/entity/number/) entity that might be a solid enough option and would work for a charging limit entity.

**Comment by alandtse** [comment]
I wouldn't want to go too far off the trail but to talk the entity approach through. Cn we tie the number entity to the max range switch? For example, any number besides 90% changes the command to the argument based setting?


If BMW is doing the exposed services, maybe we should too. We could make it generic so if you know the endpoint and the command, we pass it through without further checking. Do you want to take a shot at that?

**Comment by mattsch** [comment]
> I wouldn't want to go too far off the trail but to talk the entity approach through. Cn we tie the number entity to the max range switch? For example, any number besides 90% changes the command to the argument based setting?

Possibly? But I think the two sets of operations are really distinct things. Max/standard is the "easy" route to use the defined upstream values while setting custom is a more "advanced" method. IMO, if you're going the advanced way you probably know why and aren't too bothered about the easy being in sync.

> 
> If BMW is doing the exposed services, maybe we should too. We could make it generic so if you know the endpoint and the command, we pass it through without further checking. Do you want to take a shot at that?

I'm happy to start implementing the various services, but I would much prefer to not try a single generic option since that will be rather cumbersome from a user perspective. Thinking of hitting the basics of:

- wake
- set_valet_mode
- set_charge_limit
- share (maybe, would only be for location sharing, docs look a bit iffy so not 100% on it)
- window_control for vent/close
- combined alerts command for horn and lights triggering
- trigger_homelink

**Comment by mattsch** [comment]
I have a quick proof of concept for the heated seats using numbers and with the custom numberbox card (on the bottom) it doesn't look half bad:
<img width="496" alt="Screen Shot 2021-04-28 at 09 25 02" src="https://user-images.githubusercontent.com/134440/116411484-c2b5e180-a803-11eb-903a-76b3a44082f0.png">

It's a little weird though, the API doesn't allow setting the heated seat level if the climate control is off. But I think this approach for that and the charging level options would be better than a service. I'll get a PR in for both of those later tonight.

**Comment by Trupal00p** [comment]
My 2 cents: Not super strong feelings but either way but I do like the service approach (for the heated seats) since there isn't an entity designed specifically for this at the moment and it feels a little "shoehorned" otherwise.

The other part of this to consider is I tend to mentally think about this as off/low/medium/high (rather than 0,1,2,3) so voice commands might be smoother with a service approach. 

Also maybe we can add a hook to turn on the climate if it is off when trying to activate a seat, we have access to the car state at the time so may as well use it :)

**Comment by mattsch** [comment]
I'm really good with either approach. Upside to going the service route is we don't have to clutter entities up more with a bunch of seats... @alandtse any preferences on it?

I do think using the number entity makes sense for the charging limit, that one doesn't seem forced and fits the use case IMO.

**Comment by alandtse** [comment]
I think it depends on where you will contribute. If you're going to core, then you're more restricted on what you can do there (custom cards shouldn't be assumed to be a default). If you contribute to my [custom component fork](https://github.com/alandtse/tesla), then you have a little more discretion. Of course, I don't plan to fix the config flow authentication in core, so it'll depend on if you or someone else fixes it there too.

That said, I'd prefer services if there isn't a reasonable entity. Managing another custom component there definitely was pushback if you spawn too many entities. Your custom card looks nice. Are you doing 0-5 or 0, 20, 40, 60, 80, 100?

**Comment by mattsch** [comment]
Sorry to drop off, work got busy.

I'm just doing 0-3 on the number entity for the seats. The more I think about it the more I keep coming back to the service as being the better option. Since there's not really a great native option I think providing the service and letting users decide how they want to surface it in the UI. Hopefully I'll have time this weekend to get it all in place.

**Comment by alandtse** [comment]
No biggie. I was planning a weekend release so can try to get yours in if you think you can get it done. Let me know.

**Comment by mattsch** [comment]
@alandtse I'll have some time today to work on this. Where should I target the home-assistant bits? Looks like you may have a path forward in home-assistant/core#9939 to getting the changes back in upstream so not sure what makes the most sense here.

**Comment by alandtse** [comment]
You can target core but realize the login fix may not make it in for the May release or that if you have MFA the fix won't work for you.

**Comment by alandtse** [comment]
Let me know if you still plan on doing this. There's a request on my custom component for this and I may take from your branch but was holding off if you're close.

**Comment by mattsch** [comment]
Sorry I've ghosted, started a new job so haven't had time to wrap this up. It'll be at least a couple of weeks before I can pick it back up so feel free to grab it.

**Comment by alandtse** [comment]
@mattsch did you have a prototype branch already for HA? I'm not seeing it in your repos. I checked your custom and HA repos.

**Comment by mattsch** [comment]
@alandtse Pushed it to https://github.com/mattsch/teslajsonpy/tree/number_entities. Apparently, the bits that I had for the custom component got clobbered so those aren't there, but the heated seats code should work as is.

**Comment by alandtse** [comment]
Oh. I thought you had it farther along so it wouldn't be too much lift to carry it over. I may just wait since I don't use the seat heaters at all.

**Comment by ehendrix23** [comment]
Just reading this as I'm eagerly awaiting this functionality ... ok, maybe not too eager right now with close to 100 degree weather. :-)
Just 2 cents, but maybe using the new select entity made available in 2021.7? That would allow the integration to define the options, represent current setting, and handle changes?

**Comment by qJake** [comment]
I also like the idea of a service to change the state, so +1 for that.

As for representing the current state, which does appear to be a challenge... there is always the possibility of adding it as an attribute to the climate (or some other) entity and if HA users are interested in exposing that as a separate entity, well, that's what template sensors are for, aren't they? 🙂 

I wouldn't mind seeing these attributes on an entity, for example:

```json
{
    "SeatHeaterLeft": 0,
    "SeatHeaterRight": 3,
    "SeatHeaterRearLeft": "unavailable",
    "SeatHeaterRearCenter": "unavailable",
    "SeatHeaterRearRight": "unavailable",
}
```

I only have the front heated seats, so I wouldn't want the rear ones taking up three extra entities (disabled or not). Someone could either make entities with sensor templates, or, some new Lovelace UI stuff allows you to display attribute values directly now, so the sensor templates may not even be needed.
